### ISSUE TITLE
ucx_perftest throwing errors with multiple threads

### ISSUE THREAD

**Issue by rzambre** [openucx/ucx#2813]
I am seeing the following issue on both UCXv1.3 and the current master branch and on both x86 and aarch64. I did make sure that I configured UCX with `--enable-mt`. Here is how I am calling perftest:

Receiver:
>./ucx_perftest -t put_bw -d mlx5_0:1 -x rc_mlx5 (no difference with `-T 2` to match sender)

Sender:
>./ucx_perftest -t put_bw -d mlx5_0:1 -x rc_mlx5 -T 2 Receiver

Either the program stalls, or I get the "Aborted" message or I see the following:

```
[rzambre@gomez02 bin]$ ./ucx_perftest -t put_bw -d mlx5_0:1 -x rc_mlx5 -T 2 gomez03
[1534877620.184446] [gomez02:3085 :0]       perftest.c:1252 UCX  WARN  CPU affinity is not set (bound to 64 cpus). Performance may be impacted.
+--------------+-----------------------------+---------------------+-----------------------+
|              |       latency (usec)        |   bandwidth (MB/s)  |  message rate (msg/s) |
+--------------+---------+---------+---------+----------+----------+-----------+-----------+
| # iterations | typical | average | overall |  average |  overall |   average |   overall |
+--------------+---------+---------+---------+----------+----------+-----------+-----------+
[gomez02:3085 :0:3085] ib_mlx5_log.c:113  FATAL: Error on QP 0x359 wqe[001]: Local length (synd 0x1 vend 0x68) opcode RDMA_WRITE
[1534877620.251428] [gomez02:3085 :1]          rc_ep.c:83   UCX  WARN  ibv_destroy_qp() returned 22: Invalid argument
[gomez02:3085 :1:3088] ib_mlx5_log.c:113  FATAL: Error on QP 0x359 wqe[001]: Local length (synd 0x1 vend 0x68) opcode RDMA_WRITE

/home/rzambre/ucx/contrib/../src/uct/ib/mlx5/ib_mlx5_log.c: [ uct_ib_mlx5_completion_with_err() ]
      ...
      108     }
      109
      110     ucs_log(log_level, "Error on QP 0x%x wqe[%03d]: %s (synd 0x%x vend 0x%x) opcode %s",
==>   111             qp_num, wqe_counter, info, ecqe->syndrome, ecqe->vendor_err_synd,
      112             uct_ib_mlx5_cqe_err_opcode(ecqe));
      113     return status;
      114 }


/home/rzambre/ucx/contrib/../src/uct/ib/mlx5/ib_mlx5_log.c: [ uct_ib_mlx5_completion_with_err() ]
      ...
      108     }
      109
      110     ucs_log(log_level, "Error on QP 0x%x wqe[%03d]: %s (synd 0x%x vend 0x%x) opcode %s",
==>   111             qp_num, wqe_counter, info, ecqe->syndrome, ecqe->vendor_err_synd,
      112             uct_ib_mlx5_cqe_err_opcode(ecqe));
      113     return status;
      114 }

==== backtrace ====
 0 0x00000000000296ab uct_ib_mlx5_completion_with_err()  /home/rzambre/ucx/contrib/../src/uct/ib/mlx5/ib_mlx5_log.c:111
 1 0x000000000003712b uct_ib_mlx5_poll_cq()  /home/rzambre/ucx/contrib/../src/uct/ib/mlx5/ib_mlx5.inl:32
 2 0x000000000003712b uct_rc_mlx5_iface_progress()  /home/rzambre/ucx/contrib/../src/uct/ib/rc/accel/rc_mlx5_iface.c:86
 3 0x000000000040a002 ucs_callbackq_dispatch()  /home/rzambre/ucx/contrib/../src/ucs/datastruct/callbackq.h:208
 4 0x000000000041ffd5 uct_perf_test_runner<(ucx_perf_cmd_t)1, (ucx_perf_test_type_t)1, (uct_perf_data_layout_t)0, false>::run_stream_req_uni()  /home/rzambre/ucx/contrib/../src/uct/api/uct.h:1644
 5 0x000000000041ffd5 uct_perf_test_runner<(ucx_perf_cmd_t)1, (ucx_perf_test_type_t)1, (uct_perf_data_layout_t)0, false>::progress_requestor()  /home/rzambre/ucx/contrib/../src/tools/perf/uct_tests.cc:125
 6 0x000000000041ffd5 uct_perf_test_runner<(ucx_perf_cmd_t)1, (ucx_perf_test_type_t)1, (uct_perf_data_layout_t)0, false>::send_b()  /home/rzambre/ucx/contrib/../src/tools/perf/uct_tests.cc:275
 7 0x000000000041ffd5 uct_perf_test_runner<(ucx_perf_cmd_t)1, (ucx_perf_test_type_t)1, (uct_perf_data_layout_t)0, false>::run_stream_req_uni()  /home/rzambre/ucx/contrib/../src/tools/perf/uct_tests.cc:417
 8 0x000000000040a656 uct_perf_test_runner<(ucx_perf_cmd_t)1, (ucx_perf_test_type_t)1, (uct_perf_data_layout_t)0, false>::run()  /home/rzambre/ucx/contrib/../src/tools/perf/uct_tests.cc:510
 9 0x00000000004077ff ucx_perf_thread_run_test()  /home/rzambre/ucx/contrib/../src/tools/perf/libperf.c:1561
10 0x00000000004077ff ucx_perf_thread_spawn._omp_fn.0()  /home/rzambre/ucx/contrib/../src/tools/perf/libperf.c:1624
11 0x0000000000016435 omp_in_final()  ???:0
12 0x0000000000007e25 start_thread()  pthread_create.c:0
13 0x00000000000f834d __clone()  ???:0
===================
[1534877620.517276] [gomez02:3085 :1]     uct_tests.cc:278  UCX  ERROR Failed to send: Endpoint timeout
[1534877620.517286] [gomez02:3085 :1]     uct_tests.cc:278  UCX  ERROR Failed to send: Endpoint timeout
[1534877620.517291] [gomez02:3085 :1]     uct_tests.cc:278  UCX  ERROR Failed to send: Endpoint timeout
```
cc @shamisp

**Comment by yosefe** [comment]
@rzambre rc_mlx5 transport does not support multi-threaded worker on UCT level, need to run UCP tests (-t ucp_put_bw)
i've added PR #2827 to fail when instantiating such transport on a MT worker

**Comment by shamisp** [comment]
Shell we modify the test to create multiple workers (worker per thread ?)

**Comment by rzambre** [comment]
@yosefe I ran `-t ucp_put_bw` with `-T 2` but saw slightly slower message rates than single-thread message rates. Is this because some serialization is happening internally within UCP since UCT doesn't support multi-thread within a memory domain?

But if I write a multi-thread benchmark in which each thread gets its own resources: worker, memory domain, iface, etc. (like what @shamisp is suggesting), would it work?

**Comment by yosefe** [comment]
1. yes, there is extra lock
2. yes, but currently ucx_perftest does not do it. any chance you can fix perftest instead of writing a new MT benchmark?

**Comment by yosefe** [comment]
@jiguanglizipao was ucx compiled with --enable-mt?

**Comment by shamisp** [comment]
@yosefe since perftest is build as part of the tree i would suggest to report an error if "T" was used but the main library was compiled without --enable-mt

**Comment by jiguanglizipao** [comment]
@yosefe Yes, I use the ./contrib/configure-release-mt script.

**Comment by yosefe** [comment]
ok, we will check this and update

**Comment by jiguanglizipao** [comment]
@yosefe The environments:
Linux gorgon1 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:23:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**Comment by yosefe** [comment]
Thanks! Which OFED/MLNX_OFED/libibverbs version was used?

**Comment by rzambre** [comment]
>yes, but currently ucx_perftest does not do it. any chance you can fix perftest instead of writing a new MT benchmark?

@yosefe I am wrapping some other work at the moment. I will most likely get to this next month.

**Comment by jiguanglizipao** [comment]
@yosefe Updated above.

**Comment by jiguanglizipao** [comment]
@rzambre I tried writing a tiny benchmark with 1 worker per thread and `mt_workers_shared = true`. 
It seems correct but the performance is worse than 1 endpoint per thread under the mutex protection when thread numbers is larger than 2.

**Comment by shamisp** [comment]
If you allocate a worker per thread you should have no locks in the path...

**Comment by jiguanglizipao** [comment]
@shamisp Yes, no locks in the path.

**Comment by yosefe** [comment]
@jiguanglizipao what is the benchmark testing? can you share the code?

**Comment by jiguanglizipao** [comment]
@yosefe ucp_tag. I just keep the ucx part. It seems 1 worker per thread is better now. Maybe the performance is affected by cache locality of my deleted codes.
Code: [https://github.com/jiguanglizipao/UCXTagBench](https://github.com/jiguanglizipao/UCXTagBench)

**Comment by yosefe** [comment]
fixed by #2827
